src/Fields/interpolate.jl

```diff
@@ -54,33 +54,46 @@ end
 #### Use other methods if a more accurate interpolation is required
 ####
 
-@inline fractional_x_index(x::FT, ::Nothing, grid) where FT = one(FT)
-@inline fractional_y_index(y::FT, ::Nothing, grid) where FT = one(FT)
-@inline fractional_z_index(z::FT, ::Nothing, grid) where FT = one(FT)
+@inline fractional_x_index(x::FT, ::Nothing, grid) where FT = zero(FT)
+@inline fractional_y_index(y::FT, ::Nothing, grid) where FT = zero(FT)
+@inline fractional_z_index(z::FT, ::Nothing, grid) where FT = zero(FT)
 
-@inline fractional_x_index(x::FT, ::Center, grid) where FT = fractional_index(length(Center(), topology(grid, 1)(), size(grid, 1)), x, xnodes(grid, Center()))
-@inline fractional_y_index(y::FT, ::Center, grid) where FT = fractional_index(length(Center(), topology(grid, 2)(), size(grid, 2)), y, ynodes(grid, Center()))
-@inline fractional_z_index(z::FT, ::Center, grid) where FT = fractional_index(length(Center(), topology(grid, 3)(), size(grid, 3)), z, znodes(grid, Center()))
+@inline fractional_x_index(x::FT, ::Center, grid) where FT = fractional_index(length(Center(), topology(grid, 1)(), grid.Nx), x, xnodes(grid, Center())) - 1
+@inline fractional_y_index(y::FT, ::Center, grid) where FT = fractional_index(length(Center(), topology(grid, 2)(), grid.Ny), y, ynodes(grid, Center())) - 1
+@inline fractional_z_index(z::FT, ::Center, grid) where FT = fractional_index(length(Center(), topology(grid, 3)(), grid.Nz), z, znodes(grid, Center())) - 1
 
 @inline fractional_x_index(x::FT, ::Face, grid) where FT = fractional_index(length(Face(), topology(grid, 1)(), size(grid, 1)), x, xnodes(grid, Face())) - 1
 @inline fractional_y_index(y::FT, ::Face, grid) where FT = fractional_index(length(Face(), topology(grid, 2)(), size(grid, 2)), y, ynodes(grid, Face())) - 1
 @inline fractional_z_index(z::FT, ::Face, grid) where FT = fractional_index(length(Face(), topology(grid, 3)(), size(grid, 3)), z, znodes(grid, Face())) - 1
 
-@inline fractional_x_index(x::FT, ::Face,   grid::XRegRectilinearGrid) where FT = @inbounds FT((x - grid.xᶠᵃᵃ[1]) / grid.Δxᶠᵃᵃ)
-@inline fractional_x_index(x::FT, ::Center, grid::XRegRectilinearGrid) where FT = @inbounds FT((x - grid.xᶜᵃᵃ[1]) / grid.Δxᶜᵃᵃ)
-@inline fractional_y_index(y::FT, ::Face,   grid::YRegRectilinearGrid) where FT = @inbounds FT((y - grid.yᵃᶠᵃ[1]) / grid.Δyᵃᶠᵃ)
-@inline fractional_y_index(y::FT, ::Center, grid::YRegRectilinearGrid) where FT = @inbounds FT((y - grid.yᵃᶜᵃ[1]) / grid.Δyᵃᶜᵃ)
+const NotFlatTopologies = Union{Periodic, Bounded}
+
+const XRegNotFlat = RectilinearGrid{<:Any, <:NotFlatTopologies, <:Any, <:Any, <:Number}
+const YRegNotFlat = RectilinearGrid{<:Any, <:Any, <:NotFlatTopologies, <:Any, <:Any, <:Number}
+const ZRegNotFlat = RectilinearGrid{<:Any, <:Any, <:Any, <:NotFlatTopologies, <:Any, <:Any, <:Number}
```

Comment:
Why not replace `RectilinearGrid` with `AbstractGrid`?

Reply:
Or perhaps define a function like `isxregular` in `Grids`. That's probably better. This code should also be close to `RectilinearGrid`, its intimately tied to the specific implementatino of `RectilinearGrid` (ordering of type parameters etc)

Reply:
I'm not sure this part actually needs to worry about flat topologies here since flat directions are delt with further down (and the topology is a parameter of `AbstractGrid`), so here what we want is just the same as `XRegLatLonGrid` which are defined:
https://github.com/CliMA/Oceananigans.jl/blob/9637514fa10334e25e3cd730b5eb28ec10e7e472/src/Grids/latitude_longitude_grid.jl#L77-L80

So I will replicate for `RectilinearGrid`

Reply:
Can we avoid using these aliases at all and instead use a function `isxregular`?

The "alias" method + direct checking is annoying -- we have to update this code for new grids. The function-based approach is better.

Reply:
Can we please find a better design than using the aliases?

Reply:
Is there a performance benefit to dispatching vs ifelse? If not it would probably look nicer to have `isregular` functions in an ifelse?